lib/Sema/CSSimplify.cpp

```diff
@@ -4971,9 +4975,93 @@ bool ConstraintSystem::repairFailures(
     break;
   }
 
+  // Accept mutable pointers in the place of immutables.
+  if (implicitConversionAvailable(lhs, rhs))
+    conversionsOrFixes.push_back(
+        ConversionRestrictionKind::ImplicitConversion);
+
   return !conversionsOrFixes.empty();
 }
 
+ConstructorDecl *ConstraintSystem::implicitConversionAvailable(Type fromType, Type toType) {
```

Comment:
The implementation here would have to be more interesting, that's why we (me and @hborla) were talking about it being complex for the solver. We'd have to add a member constraint to lookup `init(implicit:)` on the `fromType` and apply found member with an argument `toType` to make sure that everything fits together correctly, so instead of doing manual lookup down here you'd have to always record an implicit conversion between types that  didn't match but are constrained on `Conversion` and down in `simplifyRestrictedConstraintImpl` implement it as member constraint + applicable function constraint.